init/cli/start.go

```diff
@@ -15,48 +15,99 @@
+Ensures the service defined by the static and custom configurations at service/bin/launcher-static.yml and
+var/conf/launcher-custom.yml is running and its outputs are redirecting to var/log/startup.log and other
+var/log/${SUB_PROCESS}-startup.log files. If successful, exits 0, otherwise exits 1 and writes an error message to
+stderr and var/log/startup.log.`,
+	Action: start,
 }
 
-func start() error {
-	if _, _, err := lib.GetProcessStatus(); err == nil {
-		// Process already running, don't restart it.
-		return nil
+func start(ctx cli.Context) (rErr error) {
+	outputFile, err := os.OpenFile(launchlib.PrimaryOutputFile, outputFileFlag, outputFileMode)
+	if err != nil {
+		return cli.WithExitCode(1, errors.Errorf("failed to create primary output file: %s",
+			launchlib.PrimaryOutputFile))
 	}
+	defer func() {
+		if cErr := outputFile.Close(); rErr == nil && cErr != nil {
+			rErr = cli.WithExitCode(1, errors.Errorf("failed to close primary output file: %s",
+				launchlib.PrimaryOutputFile))
+		}
+	}()
+	ctx.App.Stdout = outputFile
 
-	outputFile, err := os.Create(lib.OutputFile)
+	serviceStatus, err := getServiceStatus(ctx)
 	if err != nil {
-		return cli.WithExitCode(1, errors.Wrap(err, "failed to create startup log file"))
+		return logErrorAndReturnWithExitCode(ctx,
+			errors.Wrap(err, "failed to determine service status to determine what commands to run"), 1)
 	}
-
-	cmd, err := launchlib.CompileCmdsFromConfigFiles(lib.LauncherStaticFile, lib.LauncherCustomFile, outputFile)
+	pids, err := startService(serviceStatus.notRunningCmds)
 	if err != nil {
-		return cli.WithExitCode(1,
-			errors.Wrap(err, "failed to assemble command from static and custom configuration files"))
+		return logErrorAndReturnWithExitCode(ctx, errors.Wrap(err, "failed to start service"), 1)
+	}
+	for name, runningProc := range serviceStatus.runningProcs {
+		pids[name] = runningProc.Pid
+	}
+	if err := writePids(pids); err != nil {
+		return logErrorAndReturnWithExitCode(ctx,
+			errors.Wrap(err, "failed to record pids when starting service"), 1)
+	}
+	return nil
+}
+
+func startService(notRunningCmds map[string]launchlib.CmdWithOutputFileName) (map[string]int, error) {
+	pids := make(map[string]int)
+	for name, cmd := range notRunningCmds {
+		if err := startCommand(cmd); err != nil {
+			return nil, errors.Wrapf(err, "failed to start command '%s'", name)
+		}
+		pids[name] = cmd.Cmd.Process.Pid
 	}
+	return pids, nil
+}
 
-	if err := lib.StartCommand(cmd.Primary, outputFile); err != nil {
-		return cli.WithExitCode(1, errors.Wrap(err, "failed to start process"))
+func startCommand(cmd launchlib.CmdWithOutputFileName) (rErr error) {
+	stdout, err := os.OpenFile(cmd.OutputFileName, outputFileFlag, outputFileMode)
+	if err != nil {
+		return errors.Wrap(err, "failed to open output file: "+cmd.OutputFileName)
+	}
+	defer func() {
+		if cErr := stdout.Close(); rErr == nil && cErr != nil {
+			rErr = errors.Wrap(err, "failed to close output file: "+cmd.OutputFileName)
```

Comment:
Agreed. Although it's kind of goofy to log to a file "hey I couldn't close this file...obviously...because I just wrote to it..." but seems like the best thing to do.